synapse/util/async_helpers.py

```diff
@@ -541,24 +543,40 @@ def write(self, key: str) -> AsyncContextManager:
         if curr_writer:
             to_wait_on.append(curr_writer)
 
-        # We can clear the list of current readers since the new writer waits
+        # We can clear the list of current readers since `new_defer` waits
         # for them to finish.
         curr_readers.clear()
         self.key_to_current_writer[key] = new_defer
 
         @asynccontextmanager
         async def _ctx_manager() -> AsyncIterator[None]:
+            to_wait_on_defer = defer.gatherResults(to_wait_on)
             try:
-                await make_deferred_yieldable(defer.gatherResults(to_wait_on))
+                # Wait for all current readers and the latest writer to finish.
+                # May raise a `CancelledError` if the `Deferred` wrapping us is
+                # cancelled.
+                await make_deferred_yieldable(to_wait_on_defer)
                 yield
             finally:
-                with PreserveLoggingContext():
-                    new_defer.callback(None)
-                # `self.key_to_current_writer[key]` may be missing if there was another
-                # writer waiting for us and it completed entirely within the
-                # `new_defer.callback()` call above.
-                if self.key_to_current_writer.get(key) == new_defer:
-                    self.key_to_current_writer.pop(key)
+
+                def release() -> None:
+                    with PreserveLoggingContext():
+                        new_defer.callback(None)
+                    # `self.key_to_current_writer[key]` may be missing if there was another
+                    # writer waiting for us and it completed entirely within the
+                    # `new_defer.callback()` call above.
+                    if self.key_to_current_writer.get(key) == new_defer:
+                        self.key_to_current_writer.pop(key)
+
+                if to_wait_on_defer.called:
```

Comment:
could we, instead of all this, do:

```python
             to_wait_on_defer = defer.gatherResults(to_wait_on)
             to_wait_on_defer.addBoth(lambda _: release())
             await make_deferred_yieldable(to_wait_on_defer)
             yield
```

... possibly with more `stop_cancellation` ?  Or does that not work?

Reply:
`release()` must only be called after the yield, so I'm guessing you meant:
```py
try:
    await make_deferred_yieldable(stop_cancellation(to_wait_on_defer))
    yield
finally:
    to_wait_on_defer.addBoth(lambda _: release())
```
which I was originally hesitant to do because of re-entrancy in the happy path:
1. When we exit the await, it will be because we are inside a call to `to_wait_on_defer.callback(...)`.
2. During the `yield` / body of the context manager, we may or may not do a blocking await. If we block, then the `to_wait_on_defer.callback()` call gets exited, otherwise we remain within the `to_wait_on_defer.callback()` call.
3. We reach the finally block.
    If we're no longer inside the `callback()` call, `release()` will be called instantly with the current logging context.
    Otherwise if we're still inside the `callback()` call, `release()` won't be called until we next block or terminate and will have the sentinel context.

The sometimes-extra delay before `release()` is tricky to reason about, but I think the proposed code would be alright.

---------------------------------------------------------------------

As an alternative, I'm very tempted by @erikjohnston's suggestion of delaying CancelledErrors:
```py
try:
    await make_deferred_yieldable(delay_cancellation(to_wait_on_defer))
    yield
finally:
    release()  # but inline again
```
Where `delay_cancellation` would hold on to the CancelledError and only raise it once `to_wait_on_defer` resolves.

We'd get better maintainability at the cost of having the crud associated with cancelled requests lingering a little while longer, but maybe that's a worthwhile tradeoff.

Reply:
> release() must only be called after the yield, 

oh, duh. Yes, of course.

> The sometimes-extra delay before release() is tricky to reason about, but I think the proposed code would be alright.

ugh, right, yes, let's avoid that.

And yes, `delay_cancellation` looks like it might be worth exploring.

Reply:
I had a go at using a delay in cadfe0add7717341ae90b091cc7178ffe224cffe.
We now depend on #12180 of course.